### ISSUE TITLE
Add support for an offline reading experience (ePUB, PDF, etc.)

### ISSUE THREAD

**Issue by krilnon** [apple/swift-book#2]
The shipping book supports an offline reading experience with an ePUB version available on swift.org and on Apple Books. We should evaluate which formats we want to support going forward, and where in the publication pipeline we generate these artifacts.

Blocked by: https://github.com/apple/swift-docc/issues/354

rdar://71373710

**Comment by Gerzer** [comment]
I personally love that TSPL is available in the Apple Books store. It would be great if that availability can continue with this new publication pipeline!

**Comment by amartini51** [comment]
See also https://github.com/apple/swift-docc/issues/354, which likely requires much of the same underlying work in DocC.

**Comment by schwa** [comment]
A PDF of the TSPL book is almost required at this point. Losing the ability to download the TSPL seems like a major lose of usability and functionality.

**Comment by 7ombie** [comment]
I had a couple of technical issues reading the book offline.

On the (11th-gen) Kindle Paperwhite (with an eInk display), the code examples (including inline code) use the same (variable width font) as the rest of the text.

It'd be nice to add some basic syntax highlighting (using bold and italics) for these kinds of devices too.

In Apple Books, the syntax highlighting only works when the page is light. In Dark Mode, the code is all the same color. I'm not sure if this is a limitation of the software or an oversight with the formatting.

The issue with PDFs is that you cannot resize them for tablets, ebook readers or people with poor eyesight. They only really work on desktops, where EPUB (and Apple Books) still makes more sense, at least generally.

**Comment by amartini51** [comment]
@7ombie Thanks for this feedback about the ePUB.  However, that file was build using the legacy pre-DocC pipeline, which TSPL is no longer using.  (That's why there's no longer an ePUB version of TSPL available.)  When we investigate adding ePUB support to DocC, we can watch for these issues in case any of them re-occur.

**Comment by adamhill** [comment]
So are there **no** plans for an EPUB / PDF? 

If this is going to take a while, can a static HTML site be generated by DocC that would be compatible with tablets? 

How close is the current site that is generated able to be moved around as a ZIP archive and un-zipped on another computer for use?

Massive thanks to all the contributors for the work into DocC'ing the Swift Language spec!

**Comment by amartini51** [comment]
@adamhill You might have a better discussion in a GitHub issue specifically about DocC, such as https://github.com/apple/swift-docc/issues/354  (Is there a good way on GitHub to mark that issue as blocking this one?)  The work to add support for PDF, ePUB, or static HTML would all happen in the DocC project — TSPL doesn't have any of its own build system.

The output from DocC requires running a server today — you can't just move the generated files around.  It's more like a JavaScript web app that reads the documentation from a JSON file and renders content on the page.  The `--transform-for-static-hosting` flag (on by default) passed to  `docc convert` has something to do with not depending on server-side URL rewriting; it doesn't create plain non-JavaScript HTML output.

**Comment by adam-rocska** [comment]
Mind blown, that we no longer have Apple’s brilliant programming language’s Great Book in iBooks! Wth! I had notes, bookmarks, reading experience supporting my disability, had it offline, and the list goes on.

All of that is gone for good?

Given how long time ago the ditch was made, I guess I was patient enough with this rant.

Give us at least a lousy ebook or something.
If I’d know how to do it, I swear I’d do it…

**Comment by amartini51** [comment]
@adam-rocska Thanks for writing this up.  If you have time, and don't mind sharing it, what are the accessibility issues you run into with the ouput from DocC?  It sounds like https://github.com/apple/swift-docc/issues/354 (feature request for DocC to support offline experience) doesn't capture all of the problem you're running into — but you can also add your use case and issues to that bug, if there are aspects it doesn't cover.  It would be great if we can open bug(s) against DocC, in that other GitHub repo for accessibility problems.

As a workaround, you might try cloning this GitHub repository and rendering the markdown files locally.  No, it's not ideal, but maybe it's better than all the issues you're having with the DocC output.  There are a few places where TSPL uses DocC-specific markup (like term/definition lists and tables) but it should be 99% readable using a [Swift Markdown](https://github.com/apple/swift-markdown), a plain GitHub-flavored-markdown renderer, or even [cmark](https://github.com/commonmark/cmark/tree/master), since the majority of the formatting mostly just Commonmark.

**Comment by nbalasundara** [comment]
I agree it is quite annoying that PDF version has not been made available. Especially if I have previously made bookmarks in Apple Books. Anyways, I have created one! Not exactly the 'right' page numbers but it is a pdf. I created it by saving the web pages and concatenating those in Preview. 
[Swift 5.9.2 .pdf](https://github.com/apple/swift-book/files/13854753/Swift.5.9.2.pdf)

**Comment by De-Byts** [comment]
Indeed it would be nice to have a PDF format both for having it on device and to make
[Swift 5.9.2 .pdf](https://github.com/apple/swift-book/files/13854859/Swift.5.9.2.pdf)
 notes/bookmarks. I have created one for 5.9.2 with saving the website pages as pdf and concatenating the same in Preview.  I myself made notes in my previous edition of Swift which unfortunately won't carry over.

**Comment by De-Byts** [comment]
>
> . [Swift 5.9.2 .pdf](https://github.com/apple/swift-book/files/13854859/Swift.5.9.2.pdf)

One feature or annoyance is the hyperlinks go back to the site even when they don't need to(i.e., when the info could be a link to a point in the book itself).

**Comment by idevelop-app** [comment]
It would be really nice to have back the pub version, I use this reference mostly offline, thank you

**Comment by mjmsmith** [comment]
> [Swift 5.9.2 .pdf](https://github.com/apple/swift-book/files/13854859/Swift.5.9.2.pdf)

How did you find this? Wondering if there a version for 5.10.

**Comment by nbalasundara** [comment]
I created it. Maybe I should do it for 5.10 as well! On May 11, 2024, at 1:30 PM, Mark Smith ***@***.***> wrote:﻿

Swift 5.9.2 .pdf

How did you find this? Wondering if there a version for 5.10.

**Comment by mjmsmith** [comment]
Thank you so much! Don't know if it would be tiresome for .x releases, but it would be great to have a refresh when 6.0 arrives.

**Comment by nbalasundara** [comment]
Will do!On May 12, 2024, at 9:52 PM, Mark Smith ***@***.***> wrote:﻿
Thank you so much! Don't know if it would be tiresome for .x releases, but it would be great to have a refresh when 6.0 arrives.

**Comment by ekassos** [comment]
In case this helps anyone, here's the PDF version of the TSPL documentation I made for offline viewing based on 6.0.

[TSPL Swift 6.0-beta.pdf](https://github.com/user-attachments/files/15946443/TSPL.Swift.6.0-beta.pdf)

**Comment by nbalasundara** [comment]
Great, Thanks!!On Jun 23, 2024, at 6:18 PM, Evangelos Kassos ***@***.***> wrote:﻿
In case this helps anyone, here's the PDF version of the TSPL documentation I made for offline viewing based on 6.0.
TSPL Swift 6.0-beta.pdf

**Comment by mjmsmith** [comment]
> In case this helps anyone, here's the PDF version of the TSPL documentation I made for offline viewing based on 6.0.

Thank you!

**Comment by liyanage** [comment]
I've been working on an up to date PDF/ePUB version of the current Swift 6.x TSPL book that is based on the Markdown sources instead of a conversion of the rendered website, in order to provide a more full-featured PDF/ePUB reading experience. The files are attached to this comment and I will keep updating them as new versions of Swift get documented on the swift.org website.

This conversion features:

* A table of contents with working hyperlink entries
* For the PDF, a hierarchical bookmark structure for the sections in the book
* Working cross-references between parts of the book (internal references are orange, links to the web are blue)
* Rendering/style that aims to stay consistent with the web version

The conversion is based on [pandoc](https://pandoc.org) and its LaTeX/PDF/ePUB support. There are still some minor visual glitches, but overall I think it is already quite readable and useful.

Please follow up here if you see any issues, for now that's the best way to provide feedback.

I would like to thank @amartini51 for the extensive help and feedback in getting this conversion up and running.

Stable:

* [The-Swift-Programming-Language-6.3.pdf](https://github.com/user-attachments/files/30105808/The-Swift-Programming-Language-6.3.pdf)
* [The-Swift-Programming-Language-6.3.epub.zip](https://github.com/user-attachments/files/30105818/The-Swift-Programming-Language-6.3.epub.zip)

Beta:

* [The-Swift-Programming-Language-6.4b3.pdf](https://github.com/user-attachments/files/30105917/The-Swift-Programming-Language-6.4b3.pdf)
* [The-Swift-Programming-Language-6.4b3.epub.zip](https://github.com/user-attachments/files/30105914/The-Swift-Programming-Language-6.4b3.epub.zip)

<details><summary>Previous Versions</summary>
<p>

6.3 beta 3:

* [The-Swift-Programming-Language-6.3b3.pdf](https://github.com/user-attachments/files/26295004/The-Swift-Programming-Language-6.3b3.pdf)
* [The-Swift-Programming-Language-6.3b3.epub.zip](https://github.com/user-attachments/files/26295000/The-Swift-Programming-Language-6.3b3.epub.zip)

6.2.3:

* [The-Swift-Programming-Language-6.2.3.pdf](https://github.com/user-attachments/files/24202745/The-Swift-Programming-Language-6.2.3.pdf)
* [The-Swift-Programming-Language-6.2.3.epub.zip](https://github.com/user-attachments/files/24202746/The-Swift-Programming-Language-6.2.3.epub.zip)

6.2.1:

* [The-Swift-Programming-Language-6.2.1.pdf](https://github.com/user-attachments/files/23343743/The-Swift-Programming-Language-6.2.1.pdf)
* [The-Swift-Programming-Language-6.2.1.epub.zip](https://github.com/user-attachments/files/23343740/The-Swift-Programming-Language-6.2.1.epub.zip)

6.2:

* [The-Swift-Programming-Language-6.2.pdf](https://github.com/user-attachments/files/22478714/The-Swift-Programming-Language-6.2.pdf)
* [The-Swift-Programming-Language-6.2.epub.zip](https://github.com/user-attachments/files/22478707/The-Swift-Programming-Language-6.2.epub.zip)

6.2 Beta 1:

* [The-Swift-Programming-Language 6.2b1.pdf](https://github.com/user-attachments/files/20665016/The-Swift-Programming-Language.6.2b1.pdf)
* [The-Swift-Programming-Language 6.2b1.epub.zip](https://github.com/user-attachments/files/20665053/The-Swift-Programming-Language.6.2b1.epub.zip)

6.1:

* [The-Swift-Programming-Language 6.1.pdf](https://github.com/user-attachments/files/19859231/The-Swift-Programming-Language.6.1.pdf)
* [The-Swift-Programming-Language 6.1.epub.zip](https://github.com/user-attachments/files/19859229/The-Swift-Programming-Language.6.1.epub.zip)

6.0:

* [The-Swift-Programming-Language.pdf](https://github.com/user-attachments/files/17171987/The-Swift-Programming-Language.pdf)
* [The-Swift-Programming-Language.epub.zip](https://github.com/user-attachments/files/17261460/The-Swift-Programming-Language.epub.zip)
</p>
</details>

**Comment by logicito** [comment]
> I've been working on an up to date PDF/ePUB version of the current Swift 6 TSPL book that is based on the Markdown sources instead of a conversion of the rendered website, in order to provide a more full-featured PDF/ePUB reading experience. Here is the first version for anybody who is interested.
> 
> It features:
> 
> * A table of contents with working hyperlink entries
> * A hierarchical PDF bookmark structure for the sections in the book
> * Working cross-references between parts of the book (internal references are orange, links to the web are blue)
> * Rendering/style that aims to stay consistent with the web version
> 
> The conversion is based on [pandoc](https://pandoc.org) and its LaTeX/PDF/ePUB support. There are still some minor visual glitches, but overall I think it is already quite readable and useful.
> 
> Please follow up here if you see any issues, for now that's the best way to provide feedback.
> 
> I would like to thank @amartini51 for the extensive help and feedback in getting this conversion up and running.
> 
> * [The-Swift-Programming-Language.pdf](https://github.com/user-attachments/files/17171987/The-Swift-Programming-Language.pdf)
> * [The-Swift-Programming-Language.epub.zip](https://github.com/user-attachments/files/17261460/The-Swift-Programming-Language.epub.zip)

This is great, thank you for putting this together, it is appreciated

**Comment by bendietze** [comment]
Absolutely great, thank you so much

**Comment by ekassos** [comment]
I just released [swift-book-pdf](https://github.com/ekassos/swift-book-pdf/), a python package that offers a one-line command to generate a print-ready PDF version of the TSPL book. The final document follows the DocC rendering style used in [docs.swift.org](https://docs.swift.org/swift-book/documentation/the-swift-programming-language/) and retains all internal references and external links.

To support different use cases, it offers two rendering modes:
   - Digital mode with hyperlinks for cross-references between chapters and external links.
   - Print mode with page numbers accompanying cross-references between chapters and full URLs shown in footnotes for external links.

You can see a preview of both versions below. Hope this helps some folks generate new versions of the book as things get updated!

[swift_book_digital.pdf](https://github.com/user-attachments/files/19195823/swift_book_digital.pdf)

[swift_book_print.pdf](https://github.com/user-attachments/files/19195828/swift_book_print.pdf)

**Comment by liyanage** [comment]
> Thanks a lot [@ekassos](https://github.com/ekassos). i just used your tool to generate swift 6.2 pdf version
> 
> [TheSwiftProgrammingLanguage_6.2.pdf](https://github.com/user-attachments/files/20804040/TheSwiftProgrammingLanguage_6.2.pdf)

@Karim-Mohamed-Mousa note that I already update the ePUB and PDFs in [my comment above](https://github.com/swiftlang/swift-book/issues/2#issuecomment-2380375116) for every release, and I keep the old ones around.

6.2 is not final yet so mine is labeled 6.2 beta 1.

**Comment by ekassos** [comment]
@Genaro-Chris just published https://github.com/ekassos/swift-book-archive to do exactly that. The book is rebuilt every night if there are new commits in the swift-book repo.

Files are under version control, so you can track changes. Significant versions of the book posted to the docs.swift.org website (https://github.com/swiftlang/swift-book/tags) are additionally tracked as Releases: https://github.com/ekassos/swift-book-archive/releases.

Hope this helps!

**Comment by tree-frog** [comment]
> I've been working on an up to date PDF/ePUB version of the current Swift 6.x TSPL book that is based on the Markdown sources instead of a conversion of the rendered website, in order to provide a more full-featured PDF/ePUB reading experience. The files are attached to this comment and I will keep updating them as new versions of Swift get documented on the swift.org website.

I can't thank you enough for undertaking & maintaining this! It's a wonderful gift to the community.

I think it might be nice to add a cover & basic metadata to the ePub.

<img width="1190" height="1288" alt="Image" src="https://github.com/user-attachments/assets/5384081b-1115-4609-9e8f-2e9ec8b63fc6" />

I've put together the 3 previous cover variations used on older versions of the book, with the old version information whited out, in case that could be useful:
[Covers.zip](https://github.com/user-attachments/files/22452372/Covers.zip)

Historically, the book has had 3 main versions that differed in their title font style:
- Swift 1
- Swift 2 - 4*
- Swift 5

The versions distributed through the Apple Books additionally had an Apple logo (but I won't post them here, though I have all 3 versions stored locally). The old open-source book only had the latter 2 covers (since it wasn't around for Swift 1), so I created the Swift 1 version using the Apple Books release with the Apple logo removed).

*The Swift 2 - 4 version's font matches up with the final release of "Using Swift with Cocoa and Objective-C" (Swift 4.1) that Apple used to maintain in Apple Books, so that's the version I've been using.

Thanks again!

**Comment by ekassos** [comment]
Hi folks, thanks for keeping this conversation going! [`swift-book-pdf`](https://github.com/ekassos/swift-book-pdf) now supports generating EPUB editions of _The Swift Programming Language_ from the DocC source. If you've read the Swift Book editions [previously published](https://books.apple.com/us/book/the-swift-programming-language-swift-5-7/id881256329) on Apple Books through Swift 5.7, the new EPUB edition will feel instantly familiar.

1. ~~The [Swift Programming Series](https://books.apple.com/us/book-series/swift-programming-series/id1887443754) is available on Apple Books for free, with optimized editions of The Swift Programming Language from Swift 5.8 onward. I'll be updating the series as new Swift Book releases become available.~~
1. If you prefer to read the EPUB version with another e-reader, you can download PDF and EPUB editions across Swift releases from the [Swift Book Archive](https://github.com/ekassos/swift-book-archive).

> [!NOTE]
> **Quick update on the availability of the book on Apple Books (May 12, 2026)**
> 
>Unfortunately, the book will no longer be available on Apple Books for unspecified legal reasons outside my control. If anyone has relevant contacts at Apple or Apple Books Partner Support, I’d be grateful for an introduction.
> 
> I'm sorry to the more than 2,000 readers who got the book on Apple Books and are affected by this change. This was not the outcome I was hoping for.
> 
>You can download the [EPUB file](https://github.com/ekassos/swift-book-archive/releases/latest/download/swift_book.epub) from the [Swift Book Archive](https://github.com/ekassos/swift-book-archive) and open it with the Books app. The book will sync across your devices.

I'm also working on bringing back the [interactive Swift Playground](https://github.com/swiftlang/swift-book/issues/33) for the introductory A Swift Tour chapter that shipped with the older TSPL editions, and designing a similar PDF and EPUB conversion experience in Swift for any DocC documentation package that's closer to the `docc convert` action. I've seen some promising results with [swift-docc-render](https://github.com/swiftlang/swift-docc-render), WebKit, and some CSS magic.

Hopefully I will have more to share in the coming weeks. For now, here's the latest *A Swift Tour* chapter in Swift Playground form: [GuidedTour.playground.zip](https://github.com/user-attachments/files/26680374/GuidedTour.playground.zip)

Happy reading! 📖

**Comment by 1110010100** [comment]
> Hi folks, thanks for keeping this conversation going! [`swift-book-pdf`](https://github.com/ekassos/swift-book-pdf) now supports generating EPUB editions of _The Swift Programming Language_ from the DocC source. If you've read the Swift Book editions [previously published](https://books.apple.com/us/book/the-swift-programming-language-swift-5-7/id881256329) on Apple Books through Swift 5.7, the new EPUB edition will feel instantly familiar.
> 
> 1. The [Swift Programming Series](https://books.apple.com/us/book-series/swift-programming-series/id1887443754) is available on Apple Books for free, with optimized editions of The Swift Programming Language from Swift 5.8 onward. I'll be updating the series as new Swift Book releases become available.
> 2. If you prefer to read the EPUB version with another e-reader, you can download PDF and EPUB editions across Swift releases from the [Swift Book Archive](https://github.com/ekassos/swift-book-archive).
> 
> I'm also working on bringing back the [interactive Swift Playground](https://github.com/swiftlang/swift-book/issues/33) for the introductory A Swift Tour chapter that shipped with the older TSPL editions, and designing a similar PDF and EPUB conversion experience for any DocC documentation package that's closer to the `docc convert` action. I've seen some promising results with [swift-docc-render](https://github.com/swiftlang/swift-docc-render), WebKit, and some CSS magic.
> 
> Happy reading! 📖

Thank you so much, Evangele! I've been waiting for this for ages. Really, really love having this as an ePub! <3 From Berlin
